### PR TITLE
[WIP] Specific optional SessionId option to listen to with SessionHandler

### PR DESCRIPTION
**Pull request by alberthoekstra**
This PR provides another setting on the ```SessionHandlerOptions``` class so you can specify a single SessionId when registering a SessionHandler.

Couldn't find the CONTRIBUTING.md file. Please link so I can check that out.

I've made this change based on [this](https://github.com/Azure/azure-service-bus/issues/131) issue.

I'll have to look into the test coverage for the changes.

# Extended general description
So basically what I needed was a handler that handles 1 specific session id. 

I made an application that does a request for data by sending a message to a specific topic.
The response will be sent with the ```ReplyToSessionId``` value from the original message into another queue (Response queue).

So the application that needs to receive the message just has to listen to messages in the response queue with its SessionId. Registering a session handler on a Queue client doesn't allow me to filter on a specific SessionId.

Thinking that it would be a waste of resources to just create an if statement in the message handler I found out it was easy to implement it into the project.

# Providing functionality with this PR
The best way to show what is possible with this PR is by code. 
First the current way of solving the problem:
```c#
 const string SessionId = "B1A405E2-C354-4F00-8E9E-8D71BA5C30B3";
var queueClient = new QueueClient(ConnectionString, QueueName);

queueClient.RegisterSessionHandler((session, message, arg3) =>
{
    if (!session.SessionId.Equals(SessionId, StringComparison.OrdinalIgnoreCase))
        return Task.CompletedTask;

    // process data here

    return Task.CompletedTask;
}, new SessionHandlerOptions(args =>
{
    // handle error here
    return Task.CompletedTask;
}));
```

New way of solving the problem:
```c#
const string SessionId = "B1A405E2-C354-4F00-8E9E-8D71BA5C30B3";
var queueClient = new QueueClient(ConnectionString, QueueName);

queueClient.RegisterSessionHandler((session, message, arg3) =>
{
    // process data here

    return Task.CompletedTask;
}, new SessionHandlerOptions(args =>
{
    // handle error here
    return Task.CompletedTask;
})
{
    SessionId = SessionId
});
```

### CODE REVIEW

File: src/Microsoft.Azure.ServiceBus/SessionReceivePump.cs
Comment (SeanFeldman) [inline src/Microsoft.Azure.ServiceBus/SessionReceivePump.cs]: This means Session Handler will be pinned to the `SessionId` and process only one sessions for its entire lifetime. Is that the intent @alberthoekstra?

File: src/Microsoft.Azure.ServiceBus/SessionHandlerOptions.cs
Comment (SeanFeldman) [inline src/Microsoft.Azure.ServiceBus/SessionHandlerOptions.cs]: Assuming this PR goes through, I'd add a remark related to my comment about pinning to a single session.

Comment (SeanFeldman) [inline src/Microsoft.Azure.ServiceBus/SessionHandlerOptions.cs]:
```suggestion
         /// <summary>Gets or sets the optional session ID.</summary>
```

File: src/Microsoft.Azure.ServiceBus/SessionReceivePump.cs
Comment (SeanFeldman) [inline src/Microsoft.Azure.ServiceBus/SessionReceivePump.cs]: There's no need in `if/else`. `null` indicates any session id. Therefore, `sessionHandlerOptions.SessionId` can be passed for both cases.

**Review by SeanFeldman — COMMENTED** [review]
Might want to add a test to verify the behavior.

@nemakam can you review this?

File: src/Microsoft.Azure.ServiceBus/SessionReceivePump.cs
Comment (alberthoekstra) [reply src/Microsoft.Azure.ServiceBus/SessionReceivePump.cs]: Allright, I’ll Change that tomorrow.

File: src/Microsoft.Azure.ServiceBus/SessionReceivePump.cs
Comment (alberthoekstra) [reply src/Microsoft.Azure.ServiceBus/SessionReceivePump.cs]: Yes, that’s exactly what I want to happen.

Comment (alberthoekstra) [reply src/Microsoft.Azure.ServiceBus/SessionReceivePump.cs]: Changed t. Check latest commit.

File: src/Microsoft.Azure.ServiceBus/SessionReceivePump.cs
Comment (alberthoekstra) [reply src/Microsoft.Azure.ServiceBus/SessionReceivePump.cs]: Changed it. Check latest commit.

File: src/Microsoft.Azure.ServiceBus/SessionReceivePump.cs
Comment (SeanFeldman) [reply src/Microsoft.Azure.ServiceBus/SessionReceivePump.cs]: In that case I'd suggest to elaborate in more details under the [general description](https://github.com/Azure/azure-service-bus-dotnet/pull/624#issue-239293202) to provide team with an idea what this PR is trying to accomplish and how it would work.

Comment (alberthoekstra) [reply src/Microsoft.Azure.ServiceBus/SessionReceivePump.cs]: I’ve changed the description. Is this enough?

**Review by SeanFeldman — CHANGES_REQUESTED** [review]
@alberthoekstra you'll need the following:
1. Update XML doco to indicate that only one session will be handled by message handler.
2. Update public API by running ApiApprovals tests and updating approved API to include changes to `SesssionHandlerOptions`

File: src/Microsoft.Azure.ServiceBus/SessionHandlerOptions.cs
Comment (alberthoekstra) [reply src/Microsoft.Azure.ServiceBus/SessionHandlerOptions.cs]: I didn't know exactly what you meant. But I've extended the summary in the latest commit.

**Comment by alberthoekstra** [comment]
@seanfeldman can you help me out with the ApiApprovals test? The file is not getting included into my project when I’m opening the solution in VS. 

It has something to do with the target framework but I can’t figure out what the problem is exactly.

**Comment by SeanFeldman** [comment]
@alberthoekstra I can do that on your PR, but would need to have permissions to push to your repo.

**Comment by SeanFeldman** [comment]
Note: once PR #625 is merged into `dev`, this branch would need to be rebased to include the changes.

File: src/Microsoft.Azure.ServiceBus/SessionHandlerOptions.cs
Comment (alberthoekstra) [reply src/Microsoft.Azure.ServiceBus/SessionHandlerOptions.cs]: Changed it.

**Comment by alberthoekstra** [comment]
PR #625 is merged.

I've added you as Collaborator in the fork repo.

**Comment by SeanFeldman** [comment]
Public API changes approved [here](https://github.com/Azure/azure-service-bus-dotnet/pull/624/commits/b30d2f0b1cc4ae21f43e6b185d341dac84513e29)

**Review by SeanFeldman — DISMISSED** [review]

**Comment by SeanFeldman** [comment]
@alberthoekstra all looks good. 
This will be either released in the next minor or major, depending on the schedule.
Thank you.

File: src/Microsoft.Azure.ServiceBus/SessionHandlerOptions.cs
Comment (nemakam) [inline src/Microsoft.Azure.ServiceBus/SessionHandlerOptions.cs]: What happens when you set sessionId here and then set maxConcurrentSessions as well?
The other args do not hold up when you put a sessionId here.

Comment (SeanFeldman) [reply src/Microsoft.Azure.ServiceBus/SessionHandlerOptions.cs]: Good question. I've missed that... normally there's be a default (2000) or user specified number of concurrent pumps. When `SessionId` is specified (i.e. not null), it should be short circuited to one pump only imo.

**Review by nemakam — CHANGES_REQUESTED** [review]
Sorry for being blunt, but I don't think this PR is clean in terms of how we are approaching the fix. 
The sessionPump is built in a way that if there is no message received within the `sessionHandlerOptions.MessageWaitTimeout` time, we will close the session and then move on to the next session.
With you current change, it will end up closing and reconnecting the session when there are no messages.

The second problem is how does setting `sessionHandlerOptions.SessionId` interact with `maxConcurrentSession`? 

There was possible solution mentioned here - https://github.com/Azure/azure-service-bus/issues/131#issuecomment-412003895

**Comment by alberthoekstra** [comment]
Okay, sorry for the delay. But I've got it working.

In the MessageSession class, the OnMessageHandler method was overridden from the base class. It just threw an exception saying it's not supported.

I've removed that override and it now simply works like expected.

I'm sure I'm missing something because this is way too simple.
I couldn't find the reason why it threw the exception in the override, maybe one of you can remember?

**Comment by nemakam** [comment]
After fixing the above mentioend issue, @alberthoekstra , could you do a manual test once in your machine and let this Handler run for say 15 minutes when there are messages to be consumed, and another 15 minutes when there are no messages to be consumed. Check for any errors that pop up during that time.

**Comment by SeanFeldman** [comment]
@alberthoekstra it looks like you're very close to complete this PR. Are you sure you don't want to run that manual test to ensure all works?

**Comment by alberthoekstra** [comment]
The manual tests are not the only things @nemakam suggested right?
I think the change i've made should be reverted and there should be an own receive pump somewhere.

**Comment by SeanFeldman** [comment]
Marked PR as `[WIP]` until it's not and ready to be merged.

**Comment by axisc** [comment]
@alberthoekstra  we're closing this PR given that this repo has been migrated to [Azure SDK for .NET](https://github.com/Azure/azure-sdk-for-net/tree/master/sdk/servicebus/Microsoft.Azure.ServiceBus). 

If you feel that you still want to pursue this feature, please open an issue in the new repo and we can discuss this further.

**Pull request closed without merging**
